rs/boundary_node/rate_limits/canister/access_control.rs

```diff
@@ -0,0 +1,44 @@
+use candid::Principal;
+
+const FULL_ACCESS_ID: &str = "2vxsx-fae";
+const FULL_READ_ID: &str = "2vxsx-fae";
+
+pub trait ResolveAccessLevel {
+    fn get_access_level(&self) -> AccessLevel;
```

Comment:
Why not `get_access_level(&self, Principal) -> AccessLevel`? Otherwise, I notice below you create a new `AccessLevelResolver` every time you want to check someone's access level?

Reply:
it allows me to construct objects nicely in the canister method:
```
#[query(name = "get_rule_by_id")]
#[candid_method(query)]
fn get_rule_by_id(rule_id: RuleId) -> GetRuleByIdResponse {
    let caller_id = ic_cdk::api::caller();
    let response = with_state(|state| {
        let access_resolver = AccessLevelResolver::new(caller_id);
        let formatter =
            ConfidentialityFormatterFactory::new(access_resolver).create_rule_formatter();
        let fetcher = RuleFetcher::new(state, formatter);
        fetcher.fetch(rule_id)
    })?;
    Ok(response.into())
}
```
otherwise i'd need to pass `caller_id` somehow else.

Reply:
Why are we doing the ACLs so far down the call stack? For example, for `add_config` or `disclose_rules`, we should not even instantiate these `ConfigAdder` and `RulesDiscloser` objects. The earlier we can reject the call, the better.

Reply:
IMO in the canister function it's better to assemble the final object (by either objects composition or using wrappers) and perform just one single call on that object, which executes the whole logic. With this approach we can ensure best testability of the canister. I think this approach is common for canisters. So I would abstain from putting access level call within the canister method. Although it is probably indeed better to introduce wrapper structs: `WithLogger`, `WithAccessLevel`, `WithMetrics`.

Reply:
> For example, for add_config or disclose_rules, we should not even instantiate these ConfigAdder and RulesDiscloser objects.

We might want to log info before/after authorization, so we need a composable and extensible approach.

Reply:
> it allows me to construct objects nicely in the canister method:

Ah, but there's an even nicer way to achieve this:

```rs
// Initialize AccessLevelResolver (let's say call it ACCESS_LEVEL_RESOLVER)

#[query(name = "get_rule_by_id")]
#[candid_method(query)]
fn get_rule_by_id(rule_id: RuleId) -> GetRuleByIdResponse {
    let caller_id = ic_cdk::api::caller();

    // use ACCESS_LEVEL_RESOLVER with caller_id
    
    ...
}
```

I have some examples for doing this [here](https://gist.github.com/rikonor/4b4ccf1db12408c758171543383a7240#file-10_di_1-rs), [here](https://gist.github.com/rikonor/4b4ccf1db12408c758171543383a7240#file-10_di_2-rs) as well as [here](https://gist.github.com/rikonor/4b4ccf1db12408c758171543383a7240#file-13_ss_starter-rs) (last one includes stable structures).

Also, we use a similar approach throughout [this](https://github.com/dfinity/ic/blob/master/rs/boundary_node/certificate_issuance/certificate_orchestrator/src/main.rs#L631-L639) codebase.

Reply:
let me do this properly in the follow up PR